geniza/corpus/metadata_export.py

```diff
@@ -0,0 +1,221 @@
+CSV_EXPORT_PROGRESS = True
+SEP_WITHIN_CELLS = "; "
+CSV_EXPORT_FIELDS_DOC = [
+    "pgpid",
+    "url",
+    "iiif_urls",
+    "fragment_urls",
+    "shelfmark",
+    "multifragment",
+    "side",
+    "region",
+    "type",
+    "tags",
+    "description",
+    "shelfmarks_historic",
+    "languages_primary",
+    "languages_secondary",
+    "language_note",
+    "doc_date_original",
+    "doc_date_calendar",
+    "doc_date_standard",
+    "notes",
+    "needs_review",
+    "url_admin",
+    "initial_entry",
+    "last_modified",
+    "input_by",
+    "status",
+    "library",
+    "collection",
+]
+
+import csv
+import os
+
+from django.conf import settings
+from django.contrib.postgres.aggregates import ArrayAgg
+from django.contrib.sites.models import Site
+from django.db.models.query import Prefetch
+from django.http import StreamingHttpResponse
+from django.utils import timezone
+from rich.progress import track
+
+script_user = settings.SCRIPT_USERNAME
+site_domain = Site.objects.get_current().domain.rstrip("/")
+url_scheme = "https://"
+
+from geniza.common.utils import Echo, timeprint
+from geniza.corpus.models import Document, TextBlock
+
+
+class Exporter(object):
+    model = None
+    str_plural = "objects"
+    csv_fields = []
+
+    def __init__(
+        self, qset=None, progress=CSV_EXPORT_PROGRESS, sep_within_cells=SEP_WITHIN_CELLS
+    ):
+        self.qset = qset
+        self.progress = progress
+        self.sep_within_cells = sep_within_cells
+
+    def csv_filename(self):
+        return (
+            f'geniza-{self.str_plural}-{timezone.now().strftime("%Y%m%dT%H%M%S")}.csv'
+        )
+
+    def get_queryset(self):
+        return self.model.objects.metadata_prefetch() if not self.qset else self.qset
+
+    def get_export_data_dict(self, obj):
+        # @NOTE: THIS NEEDS TO BE SUBCLASSED
```

Comment:
could raise `NotImplementedError` in the base class if this is required to be extended